File: src/main/java/frc/team449/system/motor/WrappedMotor.java
package frc.team449.system.motor;

import edu.wpi.first.wpilibj.motorcontrol.MotorController;
import frc.team449.system.encoder.Encoder;
import io.github.oblarg.oblog.Loggable;
import io.github.oblarg.oblog.annotations.Log;
import org.jetbrains.annotations.NotNull;

public final class WrappedMotor implements MotorController, Loggable {
  public final @NotNull Encoder encoder;
  private final @NotNull MotorController motor;
  /** Name for logging */
  private final @NotNull String name;

  @Log private double voltage;

  public WrappedMotor(
      @NotNull String name, @NotNull MotorController motor, @NotNull Encoder encoder) {
    this.motor = motor;
    this.encoder = encoder;
    this.name = name;
  }

  @Override
  public void set(double speed) {
    motor.set(speed);
  }

  @Override
  public void setVoltage(double outputVolts) {
    motor.setVoltage(outputVolts);
    this.voltage = outputVolts;
  }

  @Log
  public double getVoltage() {
    return voltage;
  }

  @Log
  @Override
  public double get() {
    return motor.get();
  }

  public double getPosition() {
    return encoder.getPosition();
  }

  public double getVelocity() {
    return encoder.getVelocity();
  }

  @Override
  public boolean getInverted() {
    return motor.getInverted();
  }

  @Override
  public void setInverted(boolean isInverted) {
    motor.setInverted(isInverted);
  }

  @Override
  public void disable() {
    motor.disable();
  }

  @Override
  public void stopMotor() {
    motor.stopMotor();
  }

  @Override
  public String configureLogName() {
    return this.name;
  }
}

File: src/main/kotlin/frc/team449/system/motor/WrappedMotor.kt
package frc.team449.system.motor

import edu.wpi.first.wpilibj.RobotController
import edu.wpi.first.wpilibj.motorcontrol.MotorController
import frc.team449.system.encoder.Encoder
import io.github.oblarg.oblog.Loggable
import io.github.oblarg.oblog.annotations.Log

/** Our own wrapper grouping the motor controller and encoder for a motor
 *
 * @param name Name for logging
 */
class WrappedMotor(
  private val name: String,
  private val motor: MotorController,
  val encoder: Encoder
) : MotorController by motor, Loggable {
  @Log
  var voltage = 0.0
    private set

  /** Position in meters or whatever unit you set */
  var position: Double
    get() = encoder.position
    set(pos) {
      encoder.position = pos
    }

  /** Velocity in meters per second or whatever unit you set */
  val velocity: Double
    get() = encoder.velocity

  override fun setVoltage(volts: Double) {
    motor.setVoltage(volts)
    this.voltage = volts
  }

  override fun set(output: Double) {
    motor.set(output)
    this.voltage = output * RobotController.getBatteryVoltage()
  }

  override fun configureLogName() = this.name
}
